File: Python/283_Move_Zeroes.py
"""
Given an integer array nums, move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Note that you must do this in-place without making a copy of the array.
"""
# It can only be done in-place with the replacement method. 
# You need to use two pointers, one of which keeps scanning backwards to find a non-zero position, and then exchange the position with the previous pointer.
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        i = 0
        for j in range(len(nums)):
            if nums[j] != 0:
                nums[i], nums[j] = nums[j], nums[i]
                i += 1    
        

File: Java/283_Move_Zeroes.java
// Given an integer array nums, move all 0's to the end of it while maintaining the relative order of the non-zero elements.

// Note that you must do this in-place without making a copy of the array.

class Solution {
    public void moveZeroes(int[] nums) {
        if (nums.length <= 1) {
            return;
        }
        int i = 0;
        for (int j = 0; j < nums.length; ++j) {
            if (nums[j] != 0) {
                nums[i] = nums[j];
                if (i < j) {
                    nums[j] = 0;
                }
                i++;
            }
        }
    }
}